rfc/indexing-design.md

```diff
@@ -0,0 +1,256 @@
+
+After publishing an index, any user looking for content can query for the blob-index of that content. The user read the blob-index to determine what data shard needs to be retrieved, and then asks for the location of that shard.
+
+Finally the user uses the shard location, with range information from the blob index, to retrieve a specific piece of data from w3s.
+
+## Index Query
+
+A w3s client can query indexed information by specific attributes, and can filter retrieved query results.
+
+The queryable attributes are as follows.
+- Site: This lets the query ask for results where data is stored at specified site(s), i.e. query can specify "where site in [X]".
+- Shard: This lets the query ask for results for a specific shard, i.e. query can specify "where shard in [S]".
+- Slice: This lets the query ask for a specific block of data, identified my multihash with a result being the byte range within a shard containing the block of data.
+
+Query results can be filtered to select only results that the user has access to, or for data supplied by specific data providers.
+
+### Query Specification
+
+```
+match = {
+  subject=[myID, myOrgID],
+  location=spBaseURL,
+}
+results = Query(multihash="bafy...", match)
+```
+
+### Query Operation Diagram
+
+![w3up-query-diagram](w3up-query-diagram.png)
+
+## Indexing System Components
+
+The indexing components are intended for deployment on the w3up gateway. They may also be deployed to the w3up client, where they will not be shared with other clients.
+
+### W3up Index
+
+The w3up index layer is responsible for reading a query specification, and then
+1. Creating a set of IPNI queries needed to get the requested index content
+2. Sending the queries to the w3up IPNI cache
+3. Filtering returned results
+4. Packaging results for the query client.
+
+### W3up IPNI and Index Cache
+
+The w3up IPNI Cache is a cache that holds IPNI query results and sharded-dag-index data. The cache is:
+
+- Limited-size LRU. Discards the least recently used data when at storage capacity.
+- Temporary. Evicts items that have been cached for longer than the cache expiration time, regardless of last access. This allows changes to query results to be seen.
+- Negative (empty) result-aware. Caching empty results prevents more expensive IPNI queries for data that is not indexed. Negative cache entries are kept in a separate cache so that negative entries cannot evict a positive entrries due to LRU.
```

Comment:
```suggestion
- Negative (empty) result-aware. Caching empty results prevents more expensive IPNI queries for data that is not indexed. Negative cache entries are kept in a separate cache so that negative entries cannot evict positive entries due to LRU.
```

I'd also give them a significantly smaller cache time.